src/xhr-base.js

```diff
@@ -0,0 +1,404 @@
+   * @param {string|undefined} unusedResponseType The original response type used to
+   *     initiate the XHR.
+   * @return {*} The deserialized regular response.
+   * @protected
+   */
+  fromStructuredCloneable_(response, unusedResponseType) {
+    dev().assert(this.win.Response, 'Response object not present on window');
+    user().assert(isObject(response), 'Object expected: %s', response);
+    return new this.win.Response(response['body'], response['init']);
+  }
+
+  /**
+   * Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
+   * able to set a header (e.g. AMP-Source-Origin), but this will force
+   * preflight request on all CORS request.
+   * @param {!Window} win
+   * @param {string} url
+   * @return {string}
+   */
+  getCorsUrl(win, url) {
+    return getCorsUrl(this.win, url);
+  }
+
+  /**
+   * Serializes a fetch request so that it can be passed to `postMessage()`,
+   * i.e., can be cloned using the
+   * [structured clone algorithm](http://mdn.io/Structured_clone_algorithm).
+   *
+   * The request is serialized in the following way:
+   *
+   * 1. If the `init.body` is a `FormData`, set content-type header to
+   * `multipart/form-data` and transform `init.body` into an
+   * `!Array<!Array<string>>` holding the list of form entries, where each
+   * element in the array is a form entry (key-value pair) represented as a
+   * 2-element array.
+   *
+   * 2. Return a new object having properties `input` and the transformed
+   * `init`.
+   *
+   * The serialized request is assumed to be de-serialized in the following way:
+   *
+   * 1.If content-type header starts with `multipart/form-data`
+   * (case-insensitive), transform the entry array in `init.body` into a
+   * `FormData` object.
+   *
+   * 2. Pass `input` and transformed `init` to `fetch` (or the constructor of
+   * `Request`).
+   *
+   * Currently only `FormData` used in `init.body` is handled as it's the only
+   * type being used in AMP runtime that needs serialization. The `Headers` type
+   * also needs serialization, but callers should not be passing `Headers`
+   * object in `init`, as that fails `fetchPolyfill` on browsers that don't
+   * support fetch. Some serialization-needing types for `init.body` such as
+   * `ArrayBuffer` and `Blob` are already supported by the structured clone
+   * algorithm. Other serialization-needing types such as `URLSearchParams`
+   * (which is not supported in IE and Safari) and `FederatedCredentials` are
+   * not used in AMP runtime.
+   *
+   * @param {string} input The URL of the XHR to convert to structured
+   *     cloneable.
+   * @param {!FetchInitDef} init The options of the XHR to convert to structured
+   *     cloneable.
+   * @return {{input: string, init: !FetchInitDef}} The serialized structurally-
+   *     cloneable request.
+   * @private
+   */
+  toStructuredCloneable_(input, init) {
+    const newInit = Object.assign({}, init);
```

Comment:
Do we really need to clone here? Or can we just modify `init`?

Reply:
yep, because we are passed around in all the lifecycle methods so wouldn't want to modify it